### PR TITLE
Conditionally disabling Listings (Waiting on Product but technically Good to Go)

### PR DESCRIPTION
**Pull request by jeremyf**
## What type of PR is this? (check all applicable)

- [x] Feature

## Description

Prior to this commit, we were assuming everyone still had the Listing's
feature turned on.

With this commit, we move from the "implicitly on" nature of
`FeatureFlag.accessible?` to the "explicitly off" nature of
`FeatureFlag.enabled?`.  The key differentiator is that we are not
"pushing out a feature flag setting to other forems."  Listings now
become an "opt-in".

This requires a few things:

1) Ensure that DEV.to enables the `:listing_feature` before we deploy
   this branch.
2) Communicate this change to other Forem adopters and provide a
   timeline for when we'll deploy this change.
3) Documentation.  Given that we have this feature, but it will be
   disabled by default, we want to look into communicating what it is
   and how to enable it.

Further, I chose to do the least effort change on our specs.  Namely to
treat the specs as though someone fully enabled the listing feature.

## Related Tickets & Documents

This pull request is a "competetor" to #16423.

Closes forem/rfcs#291

Relates to #16420 and #16416.

## QA Instructions, Screenshots, Recordings

The instructions of #16406 don't quite apply, but are similar:

Testing this change:

1.   Start your local application (e.g. $ bin/startup)
2.   Go to http://localhost:3000/listings
3.   You should get an ActiveRecord::RecordNotFound error; because the
     application is now looking for a user or org named "listings"; in other
     words we're not routing GET "/listings" to listings#index controller
     action.
4.   Now enable the feature (e.g. $ bin/rails r \ "FeatureFlag.enable(:listing_feature)")
5.   Refresh http://localhost:3000/listings
6.   You should get a Successful response.
8.   Now disable the feature (e.g. $ bin/rails r \ "FeatureFlag.disable(:listing_feature)")
8.   Refresh http://localhost:3000/listings
9.   You should get an ActiveRecord::RecordNotFound error
10.  Goto 4

## Added/updated tests?

- [x] Yes

## [Forem core team only] How will this change be communicated?

- [x] I'm not sure how best to communicate this change and need help

### CODE REVIEW

**Review by Zhao-Andy — COMMENTED** [review]
Some thoughts:

> Ensure that DEV.to enables the :listing_feature before we deploy this branch.

I'm happy to update this. It's currently disabled right now. I'm a bit confused as to whether I can enable it before this PR -- for ex. if the feature needs to exist first, or if `FeatureFlag.enable(:some_non_instantiated_thing)` works.

> You should get an ActiveRecord::RecordNotFound error because the application is now looking for a user or org named "listings"

We should add `listings` to `reserved_words.rb`.

**Comment by jgaskins** [comment]
> I'm happy to update this. It's currently disabled right now. I'm a bit confused as to whether I can enable it before this PR -- for ex. if the feature needs to exist first, or if `FeatureFlag.enable(:some_non_instantiated_thing)` works.

@Zhao-Andy Enabling a feature flag automatically creates it as a feature if it doesn't exist. The `flipper` gem makes feature flags trivial to work with. I think we're the ones complicating it. :joy:

**Comment by Zhao-Andy** [comment]
> I think we're the ones complicating it. 😂 

lol gotcha 🙃

**Review by Zhao-Andy — APPROVED** [review]
LGTM, Travis failed though.

**Comment by nickytonline** [comment]
> LGTM, Travis failed though.

Looks like DB timeouts. I restarted the build node.

```
(called from /home/travis/build/forem/forem/vendor/bundle/ruby/3.0.0/gems/sidekiq-6.4.0/lib/sidekiq/client.rb:192:in `block in raw_push'}
2022-02-07 15:08:25 +0000 Rack app ("GET /signin=true" - (127.0.0.1)): #<ActiveRecord::QueryCanceled: PG::QueryCanceled: ERROR:  canceling statement due to statement timeout
CONTEXT:  while updating tuple (0,14) in relation "users"
```

File: config/initializers/cypress_rails.rb
Comment (citizen428) [inline config/initializers/cypress_rails.rb:26]: Do we need to put this into the initializer? Can we not just toggle the feature flag for the specs that need it? 

https://github.com/forem/forem/pull/16379

**Review by citizen428 — APPROVED** [review]

File: config/initializers/cypress_rails.rb
Comment (jeremyf) [reply config/initializers/cypress_rails.rb]: We can absolutely toggle it for the two files that need it.

**Comment by nickytonline** [comment]
There is a valid failure in the E2E tests.

```
    1) shows the sidebar on all feed views
  5 passing (47s)
  1 failing
  1) Home Feed Navigation
       shows the sidebar on all feed views:
     TestingLibraryElementError: Timed out retrying after 4000ms: Unable to find an accessible element with the role "heading" and name "Listings"
```

File: cypress/integration/seededFlows/homeFeedFlows/feedNavigation.spec.js
Comment (jeremyf) [reply cypress/integration/seededFlows/homeFeedFlows/feedNavigation.spec.js:126]: I'll remove those.

**Comment by nickytonline** [comment]
This doesn't block this PR, but I believe credits are only tied to listings so we should remove it from the member details page if the listings are disabled. cc: @Ridhwana 

image

**Comment by Ridhwana** [comment]
> This doesn't block this PR, but I believe credits are only tied to listings so we should remove the credits section from the member details page if the listings are disabled. cc: @Ridhwana
> 
> image

Nice catch @nickytonline. @jeremyf are we planning on eventually removing listings completely?

**Comment by citizen428** [comment]
@jeremyf This seems to have the necessary approvals. If it's not ready to merge, maybe update the description and set it back to draft.

**Comment by jeremyf** [comment]
@Ridhwana and @nickytonline we are planning on the Listings to be behind a feature flag.  I don't think we're planning on removing it (unless we spend effort extracting this to a plugin).

Regarding credits, that sounds like a "feature dependency management" issue.  I'm going to raise this with @amywtlin

**Comment by amywtlin** [comment]
@nickytonline Can you create an issue to hide credits when the feature flag of Listings is turned off so we don't forget about this? I don't think it is a blocker, but yes we should also hide credits when Listings is turned off.

**Review by amywtlin — CHANGES_REQUESTED** [review]
Waiting on #16698 to finish first

**Comment by nickytonline** [comment]
> @nickytonline Can you create an issue to hide credits when the feature flag of Listings is turned off so we don't forget about this? I don't think it is a blocker, but yes we should also hide credits when Listings is turned off.

@amywtlin, I just got back from vacation. Was an issue created for this?

**Comment by amywtlin** [comment]
Hey @nickytonline - I turned it into a discussion here before I finalize and turn them into issues: https://github.com/forem/forem/discussions/16749

**Review by amywtlin — APPROVED** [review]

**Pull request merged**
